httpx/_multipart.py

```diff
@@ -185,10 +188,29 @@ def render_data(self) -> typing.Iterator[bytes]:
             yield to_bytes(chunk)
             chunk = self.file.read(self.CHUNK_SIZE)
 
+    async def arender_data(self) -> typing.AsyncIterator[bytes]:
+        if not isinstance(self.file, AsyncIterable):
+            for chunk in self.render_data():
+                yield chunk
+            return
+
+        async for achunk in self.file:
+            if not isinstance(achunk, bytes):
+                raise TypeError(
+                    "Multipart file uploads must be opened in binary mode,"
+                    " not text mode."
+                )
```

Comment:
I wanted to check the type of file in `FileField.__init__`. But I think that it's impossible to check the type of the chunks in AsyncIterable at runtime without iterating. Besides, strictly speaking, nothing stops me in python from doing something like:


```python
import asyncio

async def aiter():
    yield b"bytes"
    yield "some str"

async def main():
    result = []
    async for chunk in aiter():
        result.append(chunk)
    print(result)


asyncio.run(main())
```

I know that this is a very strange example and it's unlikely that someone will do something like that. So I can check only the first chunk in a loop to ensure it is of the correct type. Something like:

```python
checked = False
async for achunk in self.file:
    if not checked and not isinstance(achunk, bytes):
        raise TypeError(
            "Multipart file uploads must be opened in binary mode,"
            " not text mode."
        )
    checked = True
    yield achunk
```

But I'm not sure whether it'll significantly impact the performance.

Reply:
Would something like this work?

```python
    file_aiter = self.file.__aiter__()
    try:
        achunk = await aiter.__anext__()
    except StopAsyncIteration:
        return

    if not isinstance(achunk, bytes):
        raise TypeError(
            "Multipart file uploads must be opened in binary mode,"
            " not text mode."
        )

    yield achunk

    async for achunk in file_aiter:
        yield achunk
```

Reply:
Yes, it would work, but it wouldn't catch the mixed string/bytes iterators I mentioned [here](https://github.com/encode/httpx/pull/3339#discussion_r1797103174). However, I've modified the code as you suggested, since I agree that handling mixed iterators is an edge/unlikely case in this situation..